feat(art): ui art babe!

Xiaorui-Huang opened the pull request:
<!-- Enter one line summarizing the PR -->
enhence UI 

### Motivation and context
<!-- What problem does this PR solve, or what bug does it fix? Reference issues using #issue_number -->

resolves #62 

### Describe your changes
<!-- What have you changed, and how does it work (if this is not obvious from the code), to help a reviewer understand your chosen approach. -->
new __str__ for ui and improve athethetics

### Tests performed
<!-- What tests did you run to validate your changes? -->
Manual UI test

### Before submitting this PR, please make sure:
 - [x] Your code builds cleanly without any errors

Xiaorui-Huang committed:
config(art): ui art babe! (just config now)

Xiaorui-Huang committed:
docs(ui): update ui docs & consistent variable name

Xiaorui-Huang committed:
feat(ui): expose score agent api and enhenced board ui

also fix use of int literal and replaced with constant macros

Xiaorui-Huang committed:
feat(ui): enhence shape ui

Xiaorui-Huang committed:
feat(ui): integrate all ui changes and refactor helper

Xiaorui-Huang committed:
refactor(ui): refactor helper and add docs

Xiaorui-Huang committed:
feat(ui): integrate all ui enhencement in game

also catches ctrl-c to exit game

Xiaorui-Huang commented:
![image](https://user-images.githubusercontent.com/42769475/182048528-b59ad8d8-688c-4222-9241-5d8028e9498c.png)
![image](https://user-images.githubusercontent.com/42769475/182048560-de4f77fd-8055-441f-9977-abfbc7b4feff.png)
![image](https://user-images.githubusercontent.com/42769475/182048579-6808b368-e4d1-4e37-8a2f-3207363ddda4.png)
![image](https://user-images.githubusercontent.com/42769475/182048586-a96c61bb-5370-405d-ab8c-1169626ee73c.png)

Xiaorui-Huang commented:
Game Over will be a random font to spice things up

Xiaorui-Huang commented:
you can force quit using ctrl-c

YANLING-H commented on woodoku/entity/woodoku_board.py:
```diff
@@ -187,9 +187,9 @@ def __str__(self) -> str:
                 row_str = f"{row // 2}  {VERTICAL}"  # print x coordinate
-                for col in range(9):
+                for col in range(BOARD_SIZE):
                     pos = "     "
```
If you want to use macro BOARD_SIZE instead of 9, you need to change the in_between() function in util.py. Otherwise, by changing BOARD_SIZE, our board is messed up.
<img width="250" alt="Screen Shot 2022-07-31 at 7 33 06 PM" src="https://user-images.githubusercontent.com/67870437/182049981-e2a77954-e947-49dc-a092-e878c5fa5e16.png">


Also, you may consider changing the padding between x coordinates and the game board, since for x > 9, the horizontal bars can't match those above. 

<img width="250" alt="Screen Shot 2022-07-31 at 7 32 35 PM" src="https://user-images.githubusercontent.com/67870437/182050050-c72888c0-02d9-49d1-9ee6-1781fb76a0fb.png">

Xiaorui-Huang replied:
No, the macro are not meant to be changed, Macros are constants, they will stay with a value of 9. The use here is simply to convey the meaning that 9 is the border size. 

and our game simply don't support any thing other than 9. That is not the game

YANLING-H replied:
Ok, it makes sense

YANLING-H commented on woodoku/ui/command_line_ui.py:
```diff
@@ -1,95 +1,107 @@
     def show_cannot_place(self) -> None:
-        print(red("You are not able to place the shape at the position you chose\n"))
+        print(red("\n❌ You cannot place the shape at the position you chose\n"))
 
```
You might consider adding red crosses before any warning or removing this red cross, otherwise, the format of warning messages is not consistent

Xiaorui-Huang replied:
My reasoning is that the  "cannot place shape error" error is too small compared to the arts. So the cross is there give it more attention. The other errors are directly after when you input something wrong, there is no need for extra attention and bloating the error messages. And the reason I didn't make the "cannot place shape error"  use art is that the the error is too long and arts have way to large of a font

YANLING-H replied:
Ok, it makes sense

YANLING-H commented:
I feel a bit distracting when seeing a large art of "score: 0" and "game over". @EdwardHaoranLee  may share any idea about it.

Xiaorui-Huang commented:
> I feel a bit distracting when seeing a large art of "score: 0" and "game over". @EdwardHaoranLee may share any idea about it.

As in distracted by what? Games over... you are meant to see what final score you got and that the game is over

YANLING-H commented:
> > I feel a bit distracting when seeing a large art of "score: 0" and "game over". @EdwardHaoranLee may share any idea about it.
> 
> As in distracted by what? Games over... you are meant to see what final score you got and that the game is over

The size of the art makes sense when running the game on a larger screen, so there is no problem anymore.

Xiaorui-Huang commented:
Important design change. Will also wait for @EdwardHaoranLee 's approval

EdwardHaoranLee commented on woodoku/entity/score_agent.py:
```diff
@@ -26,19 +28,30 @@ def calculate_winning(self, blocks: int, groups: int) -> None:
+            self.__score = self.get_score() + GROUP_POINTS
+            if self.__streak:
+                self.__score = self.get_score() + STREAK_POINTS * self.get_streak()
+            self.__streak = self.get_streak() + 1
         else:
-            self.__streaks = 0
+            self.__streak = 0
 
-        combos = groups - 1
-        if combos > 0:
-            self.__score += COMBO_POINTS * combos
+        self.__combo = groups - 1
+        if self.get_combo() > 0:
+            self.__score = self.get_score() + COMBO_POINTS * self.get_combo()
```
Why calling method instead of accessing the instance variables?

Xiaorui-Huang replied:
This is for mocking purposes. Since get is public methods, I can use them in combination with calculating winning to more easily set the private state of a score agent. Otherwise, it would be the current implementation that reserves engineer the desired score and how to set the state only with calculate winning(which can't be mocked since it's a void method)

Xiaorui-Huang replied:
I'd need these get method to implement the mocking in #61 so I'd probably will just copy/paste these method or cherry pick the commit of these changes to #61methods

Xiaorui-Huang replied:
reversing course on mocking see comment[https://github.com/EdwardHaoranLee/WoodokuLearn/pull/61#discussion_r936237242](url) here, so I will reverse and access instance variables directly again

Xiaorui-Huang committed:
refactor(ScoreAgent): revese course on mocking

the original plan was to use the new public getters to be mocked setup ScoreAgent
with the correct private variable values. but upon consideration a black
box approach would be less dependent on any implementation, so calculate
winning would not use any getter anymore, instead it will access the
variables directly.
